tpm2_pytss/utils.py

```diff
@@ -226,3 +228,129 @@ def unwrap(
+        ekalg: (TPM2_ALG.ECC or TPM2_ALG.RSA): Type of template to create, defaults to TPM2_ALG.ECC
+        session: (ESYS_TR): Optional session, for encrypt/decrypt and audit sessions, defaults to ESYS_TR.NONE
+
+    Returns:
+        The templase as a TPM2B_PUBLIC instance
+    """
+    symmetric = TPMT_SYM_DEF_OBJECT(
+        algorithm=TPM2_ALG.AES,
+        keyBits=TPMU_SYM_KEY_BITS(aes=128),
+        mode=TPMU_SYM_MODE(aes=TPM2_ALG.CFB),
+    )
+    attrs = (
+        TPMA_OBJECT.FIXEDTPM
+        | TPMA_OBJECT.FIXEDPARENT
+        | TPMA_OBJECT.SENSITIVEDATAORIGIN
+        | TPMA_OBJECT.ADMINWITHPOLICY
+        | TPMA_OBJECT.RESTRICTED
+        | TPMA_OBJECT.DECRYPT
+    )
+    policy = b"\x83q\x97gD\x84\xb3\xf8\x1a\x90\xcc\x8dF\xa5\xd7$\xfdR\xd7n\x06R\x0bd\xf2\xa1\xda\x1b3\x14i\xaa"
+    ecc_template = TPM2B_PUBLIC(
+        publicArea=TPMT_PUBLIC(
+            type=TPM2_ALG.ECC,
+            nameAlg=TPM2_ALG.SHA256,
+            objectAttributes=attrs,
+            authPolicy=policy,
+            parameters=TPMU_PUBLIC_PARMS(
+                eccDetail=TPMS_ECC_PARMS(
+                    symmetric=symmetric,
+                    scheme=TPMT_ECC_SCHEME(scheme=TPM2_ALG.NULL),
+                    curveID=TPM2_ECC.NIST_P256,
+                    kdf=TPMT_KDF_SCHEME(scheme=TPM2_ALG.NULL),
+                )
+            ),
+            unique=TPMU_PUBLIC_ID(ecc=TPMS_ECC_POINT(x=b"\x00" * 32, y=b"\x00" * 32)),
+        )
+    )
+    rsa_template = TPM2B_PUBLIC(
+        publicArea=TPMT_PUBLIC(
+            type=TPM2_ALG.RSA,
+            nameAlg=TPM2_ALG.SHA256,
+            objectAttributes=attrs,
+            authPolicy=policy,
+            parameters=TPMU_PUBLIC_PARMS(
+                rsaDetail=TPMS_RSA_PARMS(
+                    symmetric=symmetric,
+                    scheme=TPMT_RSA_SCHEME(scheme=TPM2_ALG.NULL),
+                    keyBits=2048,
+                ),
+            ),
+            unique=TPMU_PUBLIC_ID(rsa=b"\x00" * 256),
+        )
+    )
+    if ekalg == TPM2_ALG.ECC:
+        nvbase = 0x1C0000A
+        template = ecc_template
+    elif ekalg == TPM2_ALG.RSA:
+        nvbase = 0x1C00002
+        template = rsa_template
+    else:
+        raise ValueError(f"unsupported EK algorithm {ekalg}")
+
+    nvnonce = nvbase + 1
+    nvtemplate = nvbase + 2
+
+    indices = []
+    more = True
+    while more:
+        more, data = ectx.get_capability(
+            TPM2_CAP.HANDLES,
+            TPM2_HT.NV_INDEX << 24,
+            TPM2_PT_NV.INDEX_MAX,
+            session1=session,
+        )
+        handles = list(data.data.handles)
+        indices.extend(handles)
+
+    if nvtemplate in indices:
+        thandle = ectx.tr_from_tpmpublic(nvtemplate, session1=session)
+        tpub, _ = ectx.nv_read_public(thandle, session1=session)
+        left = tpub.nvPublic.dataSize
+        mtempl = b""
+        while left > 0:
+            off = tpub.nvPublic.dataSize - left
+            size = 0x300 if left > 0x300 else left
```

Comment:
What's this magic number from?

Reply:
I thought that was the minimum allowed for TPM2_PT_NV_BUFFER_MAX, but when I checked again and it's apparently 512, not 768 for the PC client profile

Reply:
> 

This value should ideally be read from the capability structure